crates/ln-dlc-node/src/dlc_wallet.rs

```diff
@@ -0,0 +1,362 @@
+            self.on_chain_wallet
+                .fee_rate_estimator
+                .get(ConfirmationTarget::Normal)
+                .as_sat_per_vb()
+        });
+
+        // Get temporarily reserved UTXOs from in-memory storage.
+        let mut reserved_outpoints = self.on_chain_wallet.locked_utxos.lock();
+
+        let utxos: Vec<LocalOutput> = {
+            let wallet = self.on_chain_wallet.bdk.read();
+            wallet.list_unspent().collect()
+        };
+
+        let utxos: Vec<&LocalOutput> = utxos
+            .iter()
+            .filter(|utxo| !reserved_outpoints.contains(&utxo.outpoint))
+            .filter(|utxo| !utxo.is_spent)
+            .collect();
+
+        // Filter out reserved and spent UTXOs to prevent double-spending attempts.
+        let candidates = utxos
+            .iter()
+            .map(|utxo| {
+                let tx_in = TxIn {
+                    previous_output: utxo.outpoint,
+                    ..Default::default()
+                };
+
+                // Inspired by `rust-bitcoin:0.30.2`.
+                let segwit_weight = {
+                    let legacy_weight = {
+                        let script_sig_size = tx_in.script_sig.len();
+                        (36 + VarInt(script_sig_size as u64).len() + script_sig_size + 4) * 4
+                    };
+
+                    legacy_weight + tx_in.witness.serialized_len()
+                };
+
+                // The 10101 wallet always generates SegWit addresses.
+                //
+                // TODO: Rework this once we use Taproot.
+                let is_witness_program = true;
+
+                Candidate::new(utxo.txout.value, segwit_weight as u32, is_witness_program)
+            })
+            .collect::<Vec<_>>();
+
+        // This is a standard base weight (without inputs or change outputs) for on-chain DLCs. We
+        // assume that this value is still correct for DLC channels.
+        let funding_tx_base_weight = 212;
+
+        let target = Target {
+            feerate: bdk_coin_select::FeeRate::from_sat_per_vb(fee_rate),
+            min_fee: 0,
+            value: amount,
+        };
+
+        let mut coin_selector = CoinSelector::new(&candidates, funding_tx_base_weight);
+
+        let dust_limit = 0;
+        let long_term_feerate = bdk_coin_select::FeeRate::default_min_relay_fee();
+
+        let change_policy = ChangePolicy::min_value_and_waste(
+            DrainWeights::default(),
+            dust_limit,
+            target.feerate,
+            long_term_feerate,
+        );
+
+        let metric = LowestFee {
+            target,
+            long_term_feerate,
+            change_policy,
+        };
+
+        coin_selector
+            .run_bnb(metric, 100_000)
+            .map_err(|e| dlc_manager::error::Error::WalletError((format!("{e:#}")).into()))?;
+
+        debug_assert!(coin_selector.is_target_met(target));
```

Comment:
Old code. I just copied this from the `bdk_coin_select` docs/examples. It's safe because a coin selection that didn't meet its target will just fail later.

But I agree that we should fail early. I can add this to https://github.com/get10101/10101/pull/2062.